---
title: "Eval Landsharc sims"
author: "Katie Lotterhos"
date: "May 24, 2016"
output: html_document
---
### Load packages and data
```{r setup, include=FALSE}
#setwd("/Users/katie/Desktop/MINOTAUReval")
source("evalsims/ComparePlot2.R")
source("evalsims/distanceFunctionsOther.R")
source("evalsims/getEmpiricalP2.R")
source("evalsims/Getdf.R")

if(!("devtools" %in% installed.packages())){install.packages("devtools", dependencies=TRUE)}
if(!("rrcovNA" %in% installed.packages())){install.packages("rrcovNA")}
if(!("qvalue" %in% installed.packages())){
  source("https://bioconductor.org/biocLite.R")
  biocLite("qvalue")
  }

library(qvalue)
library(rrcovNA)
library(devtools)
#install_github("NESCent/MINOTAUR")#, build_vignettes=TRUE)
library(MINOTAUR)

headdir <- "~/Desktop/MINOTAUReval/data/dog_full/"
outdir <- "~/Desktop/MINOTAUReval/data/dog_full_multvar_iHS_nSL_pi_TD_H12/"
#resultsdir <- "~/Desktop/MINOTAUReval/results/"
filelist <- list.files(outdir)
length(filelist)
if (!length(filelist)==25){print("Errr: missing simulation files")}

genelist <- read.csv("data/dog_QTL_genes.csv")
  # the column names in this file match the ones in the filelist
head(genelist)
col_match <- sub(".csv", "", sub("set1_full_","", filelist))
col_match %in% names(genelist)
  # should be all trues

sessionInfo()
```

### Loop through each dataset
The following chunk of code does for each dataset:
1) calculate multivariate distances based on default covariance matrix
2) compares covariance matrix estimated from just neutral loci to covariance matrix on all data and MCD estimate of covariance
3) calculate multivariate distances based on MCD covariance matrix

I got frustrated with flex_manhattan2.  First, it doesn't plot SNPs according to their genomic position; it plots them according to index.

```{r loop}
for (i in 11:11){#:length(filelist)){
  print(c(i, filelist[i]))
  infile <- paste(outdir, filelist[i], sep="")
  dat <- read.csv(infile, header=TRUE)
  head(dat)
  dat <- dat[order(dat$chr, dat$bp),]
  
  (focal_snps <- genelist[,c("chr","start","end", "SNP", col_match[i])])
  (sel_snps <- focal_snps[focal_snps[,5]>0.5,c("chr","SNP")])
  (sel_region <- focal_snps[focal_snps[,5]>0.5,c("chr","start", "end")])
  print(sel_snps)
  print(sel_region)

  library(plyr)
   chr <- levels(factor(dat$chr))
   dat$loc <- as.factor(dat$chr)
   dat$loc <-mapvalues(dat$loc, from= chr, to= 1:length(chr))
   tail(dat$loc)

  dat2 <- NULL
   for (j in seq_along(chr)){
    bob <- dat[dat$loc==j,]
    dat2 <- rbind(dat2, bob[-c((nrow(bob)-200):nrow(bob)),])
   }
  # remove ends of chromosome where maybe low recomb?
  dat <- dat2

   dat$col <- as.character("n")
   dat$bp2 <- c()
   for (j in seq_along(chr)){
     if (j %% 2 ==0){dat$col[dat$loc==j] <- rgb(0,0,0,0.5)}else{dat$col[dat$loc==j] <- rgb(0.5,0.5,0.5,0.5)}
     bob <- dat$bp[dat$loc==j] - min(dat$bp[dat$loc==j])
     dat$bp2[dat$loc==j] <- bob/max(bob)*0.95
     #hist(dat$bp[dat$loc==j])
   }
  #plot(dat$unstd_iHS, col=dat$col, pch=19, xaxt="n", bty="l", cex=0.3)
  
  #dat$bp3 <- sprintf("%09d",dat$bp2)
  #head(dat$bp3)
  dat$pos <- as.numeric(dat$loc)+dat$bp2
  
  x <- 1:nrow(dat)
  
  
  # this makes a manhattan plot and adds the known genes
  makeplot1 <- function(text, variable){
    # text is plotted on the y -axis
  plot(dat$pos, variable, col=dat$col, pch=19, xaxt="n", bty="l", cex=0.7, las=1, ylab=text, xlab="Chromosome", main=col_match[i])
    for (j in 1:nrow(sel_snps)){
     index <- which(dat$chr == sel_snps$chr[j] & dat$bp > sel_snps$SNP[j])[1]
     points(c(dat$pos[index], dat$pos[index]), c(-10,10), type="l", lwd=8, col=rgb(0,0,1,0.3))
    }
    axis(1,at=1.5:9.5,labels=chr)
  }
  
  # this makes a manhattan plot and adds the known genes
  # and highlights the robust points
  makeplot2 <- function(text, variable){
    plot(dat$pos,variable, col=dat$col, pch=19, cex=0.7, ylab=text, bty="l", xlab="Chromosome", las=1, main=col_match[i], xaxt="n")
    points(dat$pos[dat$robust==TRUE],variable[dat$robust==TRUE], pch=19, cex=0.1, col="yellow")
    for (j in 1:nrow(sel_snps)){
     index <- which(dat$chr == sel_snps$chr[j] & dat$bp > sel_snps$SNP[j])[1]
     points(c(dat$pos[index], dat$pos[index]), c(-10,10), type="l", lwd=8, col=rgb(0,0,1,0.3))
    }
      axis(1,at=1.5:9.5,labels=chr)
  }
  
  makeplot1("iHS", dat$unstd_iHS)
  makeplot1("nSL", dat$unstd_nSL)
  makeplot1(expression(pi), dat$win51_pi)
  makeplot1("Tajima's D", dat$win51_TajD)
  makeplot1("H12", dat$win51_H12)

  
  makeplot2("iHS", dat$unstd_iHS)
  makeplot2("nSL", dat$unstd_nSL)
  makeplot2(expression(pi), dat$win51_pi)
  makeplot2("Tajima's D", dat$win51_TajD)  
  makeplot2("H12", dat$win51_H12)    

  dat$Kd.default.log <- log(dat$Kd.default,10)
  dat$Kd.mcd.subset.log <- log(dat$Kd.mcd.subset,10)
  
  makeplot1("Md.default", dat$Md.default)
  makeplot1("Md.mcd.subset", dat$Md.mcd.subset)
  makeplot1("Hd.mcd.subset", dat$Hd.mcd.subset)
  makeplot1("Kd.mcd.subset.log", dat$Kd.mcd.subset.log)
  makeplot1("Nd.mcd.subset", dat$Nd.mcd.subset)   
  
  makeplot2("Md.default", dat$Md.default)
  makeplot2("Md.mcd.subset", dat$Md.mcd.subset)
  makeplot2("Hd.mcd.subset", dat$Hd.mcd.subset)
  makeplot2("Kd.mcd.subset.log", dat$Kd.mcd.subset.log)
  makeplot2("Nd.mcd.subset", dat$Nd.mcd.subset)  
  
  makeplot1("Hd.default", dat$Hd.default)
  makeplot1("Kd.default.log", dat$Kd.default.log) 
  makeplot1("Nd.default", dat$Nd.default)
  
  hist(dat$Md.default, breaks=100)
  hist(dat$Md.mcd.subset, breaks=100)
  hist(dat$Hd.default, breaks=100)
  hist(dat$Hd.mcd.subset, breaks=100)
  hist(dat$Kd.default.log, breaks=100)  
  hist(dat$Kd.mcd.subset.log, breaks=100)
  hist(dat$Nd.default, breaks=100) 
  hist(dat$Nd.mcd.subset, breaks=100) 

  
}


```
