---
title: "Dog data- calc power"
author: "Katie Lotterhos"
date: "May 24, 2016"
output: html_document
---

### Load packages and data
```{r setup, include=FALSE}

#setwd("/Users/katie/Desktop/MINOTAUReval")
source("evalsims/getEmpiricalP3_BH.R")

if(!("devtools" %in% installed.packages())){install.packages("devtools", dependencies=TRUE)}
if(!("rrcovNA" %in% installed.packages())){install.packages("rrcovNA")}
if(!("qvalue" %in% installed.packages())){
  source("https://bioconductor.org/biocLite.R")
  biocLite("qvalue")
  }

library(qvalue)
library(rrcovNA)
library(devtools)
#source("http://bioconductor.org/biocLite.R")
#biocLite("qvalue")
#library("qvalue")
#install_github("whitlock/OutFLANK")
#install_github("NESCent/MINOTAUR")#, ref="develop")#, build_vignettes=TRUE)
#library(MINOTAUR)
library(plyr)

headdir <- "~/Desktop/MINOTAUReval/data/dog_full/"
outdir <- "~/Desktop/MINOTAUReval/data/dog_full_multvar_iHS_pi_TD_H12/"
#resultsdir <- "~/Desktop/MINOTAUReval/results/"
filelist <- list.files(outdir, pattern = "set1_")
filelist2 <- list.files(outdir, pattern = "set1b_")
filelist
length(filelist)
length(filelist2)
if (!length(filelist)==25){print("Errr: missing simulation files")}
if (!length(filelist2)==25){print("Errr: missing simulation files")}

genelist <- read.csv("data/dog_QTL_genes.csv")
  # the column names in this file match the ones in the filelist
head(genelist)
col_match <- sub(".csv", "", sub("set1_full_","", filelist))
col_match %in% names(genelist)
  # should be all trues
```

### Loop through each dataset
The following chunk of code does for each dataset:
1) calculate multivariate distances based on default covariance matrix
2) compares covariance matrix estimated from just neutral loci to covariance matrix on all data and MCD estimate of covariance
3) calculate multivariate distances based on MCD covariance matrix
4) Make plots of signals

I got frustrated with flex_manhattan2.  First, it doesn't plot SNPs according to their genomic position; it plots them according to index.

```{r loop}
for (i in 1:length(filelist)){
  print("-----------------------------------")
  print("-----------------------------------")
  print("-----------------------------------")
  print(c(i, filelist[i]))
  infile <- paste(outdir, filelist[i], sep="")
  data <- read.csv(infile, header=TRUE)
  datb <- read.csv(paste(outdir, filelist2[i], sep=""))
  identical(data$rs, datb$rs)
  dim(datb)
  dat <- cbind(data, datb[,13:14])
  dim(dat)
  head(dat)
  dat <- dat[order(dat$chr, dat$bp),]
  
  (focal_snps <- genelist[,c("chr","start","end", "SNP", col_match[i])])
  (sel_snps <- focal_snps[focal_snps[,5]>0.5,c("chr","SNP", col_match[i])])
  #(sel_region <- focal_snps[focal_snps[,5]>0.5,c("chr","start", "end")])
  names(sel_snps)[3] <- "freq"
  print(sel_snps)
  #print(sel_region)

  # change chromosome to a number for analysis
   chr <- levels(factor(dat$chr))
   dat$loc <- as.factor(dat$chr)
   dat$loc <- mapvalues(dat$loc, from= chr, to= 1:length(chr))
   tail(dat$loc)

  #### remove ends of chromosome
  # dat2 <- NULL
  #  for (j in seq_along(chr)){
  #   bob <- dat[dat$loc==j,]
  #   dat2 <- rbind(dat2, bob[-c((nrow(bob)-200):nrow(bob)),])
  #  }
  # 
  # dat <- dat2

  ### Set colors for Manhattan plot
   dat$col <- as.character("n")
   dat$bp2 <- c()
   for (j in seq_along(chr)){
     if (j %% 2 ==0){dat$col[dat$loc==j] <- rgb(0,0,0,0.5)}else{dat$col[dat$loc==j] <- rgb(0.5,0.5,0.5,0.5)}
     bob <- dat$bp[dat$loc==j] - min(dat$bp[dat$loc==j])
     dat$bp2[dat$loc==j] <- bob/max(bob)*0.95
       # this is just a transformation of bp to a fraction for plotting
     #hist(dat$bp[dat$loc==j])
   }
  #plot(dat$unstd_iHS, col=dat$col, pch=19, xaxt="n", bty="l", cex=0.3)
  
  #dat$bp3 <- sprintf("%09d",dat$bp2)
  #head(dat$bp3)
  dat$pos <- as.numeric(dat$loc)+dat$bp2
  
  #x <- 1:nrow(dat)
  
  #############################
  ### Function to determine power to detect known loci
getPower <- function(column, upper=TRUE, winsize=50){
    winsizehalf <- winsize/2
    # upper=TRUE tests for outliers in the upper 95%
    # upper=FALSE tests for outliers lower 5% of the test statistic
    # 25 SNPs tested on either side of SNP

      ### Loop through each focal SNP and calculate power in a window around it
      CI_out.95 <- CI_out.90 <- CI_robust_75 <- c()
      statname <- names(dat)[column]
      print(statname)
      for (k in 1:nrow(sel_snps)){
        dat2 <- dat[which(dat$chr == sel_snps$chr[k]),]
        
        row <- which(dat2$bp >= sel_snps$SNP[k])[1]
       
         ### The window is 100 SNPs on either side, or end of chromosome
        if((row-200)<0){y<-1}else{y<-row-200}
        if((row+200)>nrow(dat2)){z<-nrow(dat2)}else{z<-row+200}
        cond <- y:z
        
        dat3 <- dat2[cond,]
        ### Get the variable of interest
        variable <- dat2[cond,column] # This is the whole window the CI is calculated from
        if(sum(is.na(variable))>length(variable)/2){
          return(out=data.frame(breed=col_match[i], statname, sel_snps, CI_out.95=NA, CI_out.90=NA))
          break
          }
        snp <- which(dat3$bp >= sel_snps$SNP[k])[1]
        variable_win <- variable[(snp-winsizehalf):(snp+winsizehalf)] # This is the window used to look for a signal
     
        if (upper){
          (CI.95 <- quantile(variable,probs = 0.95, type=1, na.rm = TRUE))
          (CI.90 <- quantile(variable,probs = 0.90, type=1, na.rm = TRUE))
          CI_out.95[k] <- sum(variable_win > CI.95)
          CI_out.90[k] <- sum(variable_win > CI.90)
        }
        if (!upper){
          CI.95 <- quantile(variable,probs = 0.05, type=1, na.rm = TRUE)
          CI.90 <- quantile(variable,probs = 0.1, type=1, na.rm = TRUE)
          CI_out.95[k] <- sum(variable_win < CI.95)
          CI_out.90[k] <- sum(variable_win < CI.90)
        }
      maintext <- paste(col_match[i],sel_snps$chr[k],"\n", sel_snps$SNP[k])
      print(maintext)
      plot(variable, pch=19, col=dat3$col, ylab=statname,
           main=maintext, xlab="")
      lines(c(snp,snp), c(-1000,1000))
      lines(c(snp-winsizehalf,snp-winsizehalf), c(-1000,1000), lty=2)
      lines(c(snp+winsizehalf,snp+winsizehalf), c(-1000,1000), lty=2)
      abline(CI.95,0, col="blue")
      abline(CI.90,0, col="blue", lty=2)
      
      plot(variable_win, pch=19, col=dat3$col, ylim=c(min(variable, na.rm=TRUE), max(variable, na.rm=TRUE)), ylab=statname,
           main=paste(col_match[i],sel_snps$chr[k],"\n", sel_snps$SNP[k]), xlab="")
      lines(c(winsizehalf,winsizehalf), c(-1000,1000))
      abline(CI.95,0, col="blue")
      abline(CI.90,0, col="blue", lty=2)
      }# end loop though focal SNPs
      
    out=data.frame(breed=col_match[i], statname, sel_snps, CI_out.95, CI_out.90, winsize)
    return(out)
    }### end function
  
  #getPower(6, winsize=21)  

  ### Get power for upper tail
  whichcols <- which(names(dat) %in% c("unstd_iHS", "win51_H12", 
                 "Md.default.raw","Hd.default.raw", 
                 "Nd.default.raw", "Md.mcd75.raw",
                  "Nd.mcd75.raw", "Hd.mcd75.raw",
                 "DCMS.default", "DCMS.mcd75", 
                 "CSS", "Md.rankp.mcd75")
  )
  
 
  a<- lapply(whichcols, getPower, winsize=21)
  head(a)
  
  whichcolsMin <- which(names(dat) %in% c("win51_TajD", "win51_pi"))
  b<-lapply(whichcolsMin, getPower, upper=FALSE, winsize=21)

  final <- c(a,b)

  for (j in 1:length(final)){
    tab <- data.frame(final[[j]])
    if(i==1 & j==1){
    write.table(tab, file="results/dogBreedsPowerB.txt", append=FALSE, row.names=FALSE, col.names=TRUE)
    }else{
    write.table(tab, file="results/dogBreedsPowerB.txt", append=TRUE, row.names=FALSE, col.names=FALSE)
    }
  }#end loops through j
} #end loop through filelist

```

### Session info
```{r}
sessionInfo()
```
