---
title: "Eval Landsharc sims"
author: "Katie Lotterhos"
date: "May 24, 2016"
output: html_document
---

### Load packages and data
```{r setup, include=FALSE}
makeplots=FALSE

#setwd("/Users/katie/Desktop/MINOTAUReval")
source("evalsims/ComparePlot2.R")
source("evalsims/distanceFunctionsOther.R")
source("evalsims/getEmpiricalP3_BH.R")
source("evalsims/Getdf.R")

if(!("devtools" %in% installed.packages())){install.packages("devtools", dependencies=TRUE)}
if(!("rrcovNA" %in% installed.packages())){install.packages("rrcovNA")}
if(!("qvalue" %in% installed.packages())){
  source("https://bioconductor.org/biocLite.R")
  biocLite("qvalue")
  }

library(qvalue)
library(rrcovNA)
library(devtools)
#install_github("NESCent/MINOTAUR")#, build_vignettes=TRUE)
library(MINOTAUR)
library(plyr)

headdir <- "~/Desktop/MINOTAUReval/data/dog_full/"
outdir <- "~/Desktop/MINOTAUReval/data/dog_full_multvar_iHS_pi_TD_H12/"
#resultsdir <- "~/Desktop/MINOTAUReval/results/"
filelist <- list.files(outdir, pattern = "set1")
filelist
length(filelist)
if (!length(filelist)==25){print("Errr: missing simulation files")}

genelist <- read.csv("data/dog_QTL_genes.csv")
  # the column names in this file match the ones in the filelist
head(genelist)
col_match <- sub(".csv", "", sub("set1_full_","", filelist))
col_match %in% names(genelist)
  # should be all trues

sessionInfo()
```

### Loop through each dataset
The following chunk of code does for each dataset:
1) calculate multivariate distances based on default covariance matrix
2) compares covariance matrix estimated from just neutral loci to covariance matrix on all data and MCD estimate of covariance
3) calculate multivariate distances based on MCD covariance matrix
4) Make plots of signals

I got frustrated with flex_manhattan2.  First, it doesn't plot SNPs according to their genomic position; it plots them according to index.

```{r loop}
for (i in 1:length(filelist)){
  print("-----------------------------------")
  print("-----------------------------------")
  print("-----------------------------------")
  print(c(i, filelist[i]))
  infile <- paste(outdir, filelist[i], sep="")
  dat <- read.csv(infile, header=TRUE)
  dim(dat)
  head(dat)
  dat <- dat[order(dat$chr, dat$bp),]
  
  (focal_snps <- genelist[,c("chr","start","end", "SNP", col_match[i])])
  (sel_snps <- focal_snps[focal_snps[,5]>0.5,c("chr","SNP")])
  #(sel_region <- focal_snps[focal_snps[,5]>0.5,c("chr","start", "end")])
  print(sel_snps)
  #print(sel_region)

  # change chromosome to a number for analysis
   chr <- levels(factor(dat$chr))
   dat$loc <- as.factor(dat$chr)
   dat$loc <- mapvalues(dat$loc, from= chr, to= 1:length(chr))
   tail(dat$loc)

  #### remove ends of chromosome
  # dat2 <- NULL
  #  for (j in seq_along(chr)){
  #   bob <- dat[dat$loc==j,]
  #   dat2 <- rbind(dat2, bob[-c((nrow(bob)-200):nrow(bob)),])
  #  }
  # 
  # dat <- dat2

  ### Set colors for Manhattan plot
   dat$col <- as.character("n")
   dat$bp2 <- c()
   for (j in seq_along(chr)){
     if (j %% 2 ==0){dat$col[dat$loc==j] <- rgb(0,0,0,0.5)}else{dat$col[dat$loc==j] <- rgb(0.5,0.5,0.5,0.5)}
     bob <- dat$bp[dat$loc==j] - min(dat$bp[dat$loc==j])
     dat$bp2[dat$loc==j] <- bob/max(bob)*0.95
       # this is just a transformation of bp to a fraction for plotting
     #hist(dat$bp[dat$loc==j])
   }
  #plot(dat$unstd_iHS, col=dat$col, pch=19, xaxt="n", bty="l", cex=0.3)
  
  #dat$bp3 <- sprintf("%09d",dat$bp2)
  #head(dat$bp3)
  dat$pos <- as.numeric(dat$loc)+dat$bp2
  
  x <- 1:nrow(dat)
  
  ### Function to determine power to detect known loci
getPower <- function(column, upper=TRUE, nsnps=1){
    # upper=TRUE tests for outliers in the upper 95%
    # upper=FALSE tests for outliers lower 5% of the test statistic
    # 25 SNPs tested on either side of SNP
    variable <- dat[,column]
    print(names(dat)[column])
    if (upper){
      (CI.95 <- quantile(variable,probs = 0.95, type=1))
      (CI.90 <- quantile(variable,probs = 0.90, type=1))
    }
    if (!upper){
      CI.95 <- quantile(variable,probs = 0.05, type=1)
      CI.90 <- quantile(variable,probs = 0.1, type=1)
    }
    
    row <- CI_out.95 <- CI_out.90 <- CI_robust_75 <- CI_robust_99 <- c()
    for (k in 1:nrow(sel_snps)){
        row[k] <- which(dat$chr == sel_snps$chr[k] 
                        & dat$bp > sel_snps$SNP[k])[1]
        cond <- (row[k]-25):(row[k]+25)
        
        if(upper){
          CI_out.95[k] <- sum((variable[cond]) > CI.95)
          CI_out.90[k] <- sum((variable[cond]) > CI.90)
          var_75 <- max(variable[dat$robust75])
          var_99 <- max(variable[dat$robust99])
            # at least one SNP in window has a signal
            # greater than the criteria
          CI_robust_75[k] <- sum((variable[cond]) >
                                   var_75) 
          CI_robust_99[k] <- sum((variable[cond]) >
                                   var_99) 
        }# end if upper

        if (!upper){
          CI_out.95[k] <- sum((variable[cond]) < CI.95)
          CI_out.90[k] <- sum((variable[cond]) < CI.90)
                # at least five SNPs in window has a signal
                # greater than the criteria
          var_75 <- min(variable[dat$robust75])
          var_99 <- min(variable[dat$robust99])
          CI_robust_75[k] <- sum((variable[cond]) <
                                   var_75)
          CI_robust_99[k] <- sum((variable[cond]) <
                                      var_99)
        }# end if !upper
    }# end loop though focal SNPs
    out=data.frame(CI_out.95, CI_out.90, CI_robust_75, CI_robust_99)
    #print(out)
    power = (colSums(out>=nsnps))/nrow(out)
    power.out = data.frame(cutoff.CI.95=CI.95,
                           cutoff.CI.90=CI.90,
                           cutoff.robust75=var_75, 
                     cutoff.robust99=var_99,
                     power.CI.95=power[1],
                     power.CI.90=power[2],
                      power.robust75=power[3],
                      power.robust99=power[4]
    )
    return(power.out)
  }
  ### End function
  
  ### Get power
  whichcols <- which(names(dat) %in% c("unstd_iHS", "win51_H12", 
                 "pcs.default","Hcd.default", 
                 "Md.default", "Hd.default",
                 "Kd.default", "Nd.default", 
                 "Md.mcd.subset.75", "Hd.mcd.subset.75", 
                 "Kd.mcd.subset.75", "Nd.mcd.subset.75",
                 "Md.mcd.subset.99", "Hd.mcd.subset.99", 
                 "Kd.mcd.subset.99", "Nd.mcd.subset.99")
  )
  
for (nsnps in 1:5){  
  a<- lapply(whichcols, getPower, nsnps=nsnps)
  a<-data.frame(matrix(unlist(a), ncol = 8, byrow=TRUE))
  names(a) <- names(getPower(7))
  a$stat <- names(dat)[whichcols]
  a$upper<-TRUE
  a$nsnps<-nsnps
  a
  
  whichcolsMin <- which(names(dat) %in% c("win51_TajD", "win51_pi"))
  b<-lapply(whichcolsMin, getPower, upper=FALSE, nsnps=nsnps)
  b<-data.frame(matrix(unlist(b), ncol = 8, byrow=TRUE))
  names(b) <- names(getPower(7))
  b$stat <- names(dat)[whichcolsMin]
  b$upper <- FALSE
  b$nsnps<-nsnps
  final <- rbind(a,b)
  
  final$breed <- col_match[i]
  
  if(i==1 & nsnps==1){
    write.table(final, file="results/dogBreedsPower.txt", append=FALSE, row.names=FALSE, col.names=TRUE)
  }else{
    write.table(final, file="results/dogBreedsPower.txt", append=TRUE, row.names=FALSE, col.names=FALSE)
  }
}#end loop through nsnps
} #end loop
  
  ### Plotting ###
  ################
if(makeplots==TRUE){
  ### this makes a manhattan plot and adds the known genes
  makeplot1 <- function(text, variable){
    # text is plotted on the y -axis
  plot(dat$pos, variable, col=dat$col, pch=19, xaxt="n", bty="l", cex=0.7, las=1, ylab=text, xlab="Chromosome", main=col_match[i])
    for (j in 1:nrow(sel_snps)){
     index <- which(dat$chr == sel_snps$chr[j] & dat$bp > sel_snps$SNP[j])[1]
     points(c(dat$pos[index], dat$pos[index]), c(-10,10), type="l", lwd=8, col=rgb(0,0,1,0.3))
    }
    axis(1,at=1.5:9.5,labels=chr)
  }
  
  ### this makes a manhattan plot and adds the known genes
  ### and highlights the robust points
  makeplot2 <- function(text, variable, whichrobust){
    plot(dat$pos,variable, col=dat$col, pch=19, cex=0.7, ylab=text, bty="l", xlab="Chromosome", las=1, main=col_match[i], xaxt="n")
    points(dat$pos[whichrobust],variable[whichrobust], pch=19, cex=0.1, col="yellow")
    for (j in 1:nrow(sel_snps)){
     index <- which(dat$chr == sel_snps$chr[j] & dat$bp > sel_snps$SNP[j])[1]
     points(c(dat$pos[index], dat$pos[index]), c(-10,10), type="l", lwd=8, col=rgb(0,0,1,0.3))
    }
      axis(1,at=1.5:9.5,labels=chr)
  }
  
  make3 <- function(nombre, ko){
    makeplot1(nombre, ko)
    makeplot2(paste(nombre, "robust 75"), ko, which(dat$robust75))
    makeplot2(paste(nombre, "robust 99"), ko, which(dat$robust99))
  }
  
  dir.create(paste("results/dogbreeds/", col_match[i],"/", sep=""))
  png(paste("results/dogbreeds/", col_match[i],"/",col_match[i], "%02d.png", sep=""),
      width=8, height=3, units="in", res=300)
  make3("iHS", dat$unstd_iHS)
  make3(expression(pi), dat$win51_pi)
  make3("Tajima's D", dat$win51_TajD)
  make3("H12", dat$win51_H12)
  
  make3("Md.default", dat$Md.default)
  make3("Md.mcd.subset.75", dat$Md.mcd.subset.75)
  make3("Md.mcd.subset.99", dat$Md.mcd.subset.99)

  make3("Hd.default", dat$Hd.default)
  make3("Hd.mcd.subset.75", dat$Hd.mcd.subset.75)
  make3("Hd.mcd.subset.99", dat$Hd.mcd.subset.99)
  
  make3("log(Kd.default)", log(dat$Hd.default,10))
  make3("log(Kd.mcd.subset.75)", log(dat$Hd.mcd.subset.75,10))
  make3("log(Kd.mcd.subset.99)", log(dat$Hd.mcd.subset.99,10))  
  
  make3("Nd.default", dat$Nd.default)
  make3("Nd.mcd.subset.75", dat$Nd.mcd.subset.75)
  make3("Nd.mcd.subset.99", dat$Nd.mcd.subset.99)
  
  dev.off()
  # hist(dat$Md.default, breaks=100)
  # hist(dat$Md.mcd.subset, breaks=100)
  # hist(dat$Hd.default, breaks=100)
  # hist(dat$Hd.mcd.subset, breaks=100)
  # hist(dat$Kd.default.log, breaks=100)  
  # hist(dat$Kd.mcd.subset.log, breaks=100)
  # hist(dat$Nd.default, breaks=100) 
  # hist(dat$Nd.mcd.subset, breaks=100) 
  }#end if makeplots
}


```
